### DISCUSSION TITLE
Naming: how to name method overloads which return nullable result?

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/runtime#87752] [Q&A]
In LINQ, we have `First` and `FirstOrDefault`, etc., where `First` throws on empty collection and `FirstOrDefault` returns null. I actually created 2 other extensions, `FirstOrNull` and `FirstOrNullStruct`, because I don't like the "default" semantics and prefer "real" `T?` for value types, so I get `(int?)null` instead of `(int)0`.
In other places, I have other methods, like `IConfiguration.GetResolvedValueRequired` and `IConfiguration.GetResolvedValue`, which access config settings (and do some other logic on top of the regular `GetValue`), and the "required" one throws, and the other one returns null, if the config setting is missing.
Yes in other places, I have methods like `IContainer.Resolve` and `IContainer.ResolveOptional` (for dependency injection), where `Resolve` throws and `ResolveOptional` returns null.
Today, I realized that I have 3 different naming schemes for essentially the same pattern.
I was curious if other members encountered this naming challenge, what they went with, - any thoughts on this subject would be welcome. Thanks in advance.

PS. I'm currently leaning towards standardizing on the `Foo` and `FooOrNull` naming scheme, but wanted to hear other opinions before I commit to it.

**Comment by davidfowl** [comment]
I think it depends on what method existed first and if nullability was designed into the API from the get go. Personally, I've been using GetFoo and GetRequiredFoo.

**Comment by En3Tho** [comment]
Sometimes I just use Try* prefix. Yeah it's kinda occupied by bool + out T pattern but pattern matching wise it's much nicer to return T?

Also it's somewhat inspired by F# where Try* thingies return Option of T

The problem is that if you want to be consistent with api that is already there you kinda have to adhere to the naming used there or else it will look ugly. 

So maybe all of these are good options.

**Comment by RenderMichael** [comment]
When it comes to the `FirstOrNull` and `FirstOrNullStruct` methods, see https://github.com/dotnet/csharplang/discussions/6308 and add a reaction if you're in support of how it would simplify your naming convention

**Reply by TahirAhmadov** [reply]
Thank you, I already saw and voted up that proposal.
